### PR TITLE
Tessellation neighbors

### PR DESCRIPTION
**Pull request by samwaseda**
I added two algorithms to find neighbouring atoms:

- Voronoi tessellation: Indices of atoms which share the same Voronoi vertices
- Delaunay tessellation: Indices of atoms which form tetrahedrons in which no atom exists

To be honest, I don't really know whether the Delaunay tessellation finds its utility, but since the functionality was nearly the same I implemented it anyway.

### CODE REVIEW

**Comment by samwaseda** [comment]
Sorry there's an error

**Review by pmrv — APPROVED** [review]

File: tests/atomistics/structure/test_analyse.py [hunk 1]
```diff
@@ -174,6 +174,17 @@ def test_strain(self):
         self.assertLess(np.absolute(eps_yz-strain[:, 1, 2]).max(), 0.01)
         self.assertLess(np.absolute(eps_xz-strain[:, 0, 2]).max(), 0.01)
 
+    def test_tessellations(self):
```
Comment (liamhuber) [inline tests/atomistics/structure/test_analyse.py:177]: Please add `msg` kwarg to the tests to clarify the intent for skimmers.

File: pyiron_atomistics/atomistics/structure/analyse.py [hunk 2]
```diff
@@ -613,3 +616,42 @@ def get_strain(self, ref_structure, num_neighbors=None, only_bulk_type=False):
...
+        Returns:
+            pairs (ndarray): Pair indices
+        """
+        positions, indices = self._structure.get_extended_positions(
+            width_buffer, return_indices=True
+        )
+        voro = Voronoi(positions)
+        x = positions[voro.ridge_points]
+        return indices[voro.ridge_points[
+            np.isclose(self._structure.get_wrapped_coordinates(x), x).all(axis=-1).any(axis=-1)
```
Comment (liamhuber) [inline pyiron_atomistics/atomistics/structure/analyse.py]: This is beautful but kind of opaque...could we extract it to a little named method?

**Review by liamhuber — COMMENTED** [review]
Yeah, super. A couple of nits and a refactor (I'll make a PR into this branch) but otherwise lgtm.

File: pyiron_atomistics/atomistics/structure/analyse.py [hunk 3]
```diff
@@ -689,3 +689,63 @@ def get_delaunay_neighbors(self, width_buffer: float = 10.) -> np.ndarray:
...
+        coo = coo_matrix((labels, (np.arange(len(labels)), indices)))
+        labels = coo.max(axis=0).toarray().flatten()
+        # make labels look nicer
+        labels = np.unique(labels, return_inverse=True)[1]
+        mean_positions = get_mean_positions(
+            positions, self._structure.cell, self._structure.pbc, labels
+        )
+        if return_labels:
+            return mean_positions, labels
+        return mean_positions
```
Comment (liamhuber) [inline pyiron_atomistics/atomistics/structure/analyse.py]: PEP8 nit: newline at end

File: tests/atomistics/structure/test_analyse.py [hunk 4]
```diff
@@ -185,6 +185,16 @@ def test_tessellations(self):
...
+        self.assertEqual(len(bulk.analyse.cluster_positions(positions)), len(bulk))
+        self.assertEqual(
+            bulk.analyse.cluster_positions(np.zeros((2, 3)), return_labels=True)[1].tolist(),
+            [0, 0]
+        )
+
 
 if __name__ == "__main__":
-    unittest.main()
+    unittest.main()
```
Comment (liamhuber) [inline tests/atomistics/structure/test_analyse.py:200]: We actually no longer (never?) need this whole clause at the end. The CI works fine without it, and I've never run into trouble running the tests locally with out it either. I'd just cut it.

File: pyiron_atomistics/atomistics/structure/analyse.py [hunk 5]
```diff
@@ -689,3 +689,63 @@ def get_delaunay_neighbors(self, width_buffer: float = 10.) -> np.ndarray:
...
+                together across PBC are missed. Default: Same as eps
+            return_labels (bool): Whether to return the labels given according to the grouping
+                together with the mean positions
+
+        Returns:
+            positions (numpy.ndarray): Mean positions
+            label (numpy.ndarray): Labels of the positions (returned when `return_labels = True`)
+        """
+        if positions is None:
+            positions = self._structure.positions
```
Comment (liamhuber) [inline pyiron_atomistics/atomistics/structure/analyse.py]:
```suggestion
        positions = self._structure.positions if positions is None else np.array(positions)
```

I guess the structure.positions are already safely an array so we can one-line this no problem

File: pyiron_atomistics/atomistics/structure/analyse.py [hunk 6]
```diff
@@ -689,3 +689,63 @@ def get_delaunay_neighbors(self, width_buffer: float = 10.) -> np.ndarray:
...
+        positions = np.array(positions)
+        if buffer_width is None:
+            buffer_width = eps
```
Comment (liamhuber) [inline pyiron_atomistics/atomistics/structure/analyse.py]:
```suggestion
        buffer_width = eps if buffer_width is None else buffer_width
```

I'm really quite smitten with these one-lines for None cases, but ignore it if you don't like it.

**Review by liamhuber — APPROVED** [review]
A couple nits, but looks good.

Today is my first day of parental leave and then I'll be moving in January. I hope to be a but more involved again come March/April, but until then I'll probably only poke my nose into architecture stuff on base. Just a heads-up that future reviews from me might be tardy on the scale of months 😉

File: tests/atomistics/structure/test_analyse.py (hunk 4 above)
Comment (samwaseda) [reply tests/atomistics/structure/test_analyse.py:200]: otherwise it doesn't work on my terminal

Comment (jan-janssen) [reply tests/atomistics/structure/test_analyse.py:200]: I agree with @samwaseda We should have it in every test class to be compatible to executing the tests on the command line.

**Pull request merged**